Fabric.Authorization.API/scripts/AzureAD-Migration.ps1

```diff
@@ -0,0 +1,192 @@
+}
+Import-Module -Name $fabricInstallUtilities -Force
+
+function Get-NonMigratedActiveDirectoryGroups {
+    param(
+        [Parameter(Mandatory=$true)]
+        [string] $connectionString
+    )
+
+    $connection = New-Object System.Data.SqlClient.SqlConnection($connectionString)	
+    $sql = "SELECT GroupId, Name, IdentityProvider, ExternalIdentifier, TenantId
+              FROM Groups
+              WHERE IsDeleted = 0
+              AND IdentityProvider = 'Windows'
+              AND Source = 'Directory'"
+    $command = New-Object System.Data.SqlClient.SqlCommand($sql, $connection)
+    
+    Write-DosMessage -Level "Information" -Message "Retrieving groups for migration..."
+
+    $groups = @();
+    try {	
+        $connection.Open()
+        $adapter = New-Object System.Data.sqlclient.sqlDataAdapter $command
+        $groups = New-Object System.Data.DataTable
+        $adapter.Fill($groups) | Out-Null
+        $connection.Close()
+    }
+    catch [System.Data.SqlClient.SqlException] {
+        Write-DosMessage -Level "Fatal" -Message "An error occurred while executing the command to retrieve non-migrated AD groups. Connection String: $($connectionString). Error $($_.Exception)"
+    }    	
+    
+    Write-DosMessage -Level Information -Message "$($groups.Rows.Count) group(s) found for migration"
+    return $groups;
+}
+
+function Get-AzureADGroupBySID {
+    param(
+        [Parameter(Mandatory=$true)]
+        [string] $tenantId,
+        [Parameter(Mandatory=$true)]
+        [string] $groupSID
+    )
+
+    $azureADGroups = @()
+    try {
+        # connect to Azure AD
+        Connect-AzureADTenant -tenantId $tenantId
+
+        Write-DosMessage -Level "Information" -Message "Retrieving group $($groupSID) from Azure AD Tenant $($tenantId)..."
+        $azureADGroups = Get-AzureADGroup -Filter "onPremisesSecurityIdentifier eq '$($groupSID)'"
+
+        # disconnect from Azure AD
+        Disconnect-AzureAD
+    }
+    catch {
+        Write-DosMessage -Level "Fatal" -Message "Error while attempting to retrieve Azure AD group $($groupSID): $($_.Exception)"
+    }
+
+    if ($azureADGroups.Count -eq 1) {
+        return $azureADGroups[0]
+    }
+    elseif ($azureADGroups.Count -eq 0) {
+        $errorMsg = "No match found for SID $($groupSID) in Azure AD."
+        Write-DosMessage -Level "Error" -Message $errorMsg
+        throw $errorMsg
+    }
+    else {
+        $errorMsg = "Multiple matches found for group SID $($groupSID)."
+        Write-DosMessage -Level "Error" -Message $errorMsg
+        throw $errorMsg
+    }
+}
+
+function Move-ActiveDirectoryGroupsToAzureAD {
+     param(
+        [Parameter(Mandatory=$true)]
+        [string] $connString
+    )
+
+    Write-DosMessage -Level "Information" -Message "Migrating AD groups to Azure AD..."
+
+    $allowedTenantIds = Get-AzureADTenants -installConfigPath $installConfigPath
```

Comment:
Should this call have a try/catch?

Reply:
All it does is pull settings from a config file. If it fails, we should let the exception bubble up.

Reply:
I suppose we could add a try/catch and log the error and then re-throw.